### PR TITLE
[16.0][ADD] account_analytic_tag

### PR DESCRIPTION
**Pull request by Yadier-Tecnativa**
Restore account.analytic.tag
Solve the issue https://github.com/OCA/account-analytic/issues/551

@Tecnativa TT44062

### CODE REVIEW

**Comment by Yadier-Tecnativa** [comment]
ping @sergio-teruel , @pedrobaeza

File: account_analytic_tag/models/account_analytic_tag.py
Comment (pedrobaeza) [inline account_analytic_tag/models/account_analytic_tag.py]: This one is not needed anymore. You only need the tag

File: account_analytic_tag/models/account_analytic_tag.py
Comment (pedrobaeza) [inline account_analytic_tag/models/account_analytic_tag.py]: Remove these 2 fields. We are only wanting the tag part.

File: account_analytic_tag/models/account_analytic_tag.py
Comment (pedrobaeza) [inline account_analytic_tag/models/account_analytic_tag.py]: This should go into a separate file.

File: account_analytic_tag/readme/DESCRIPTION.rst
Comment (pedrobaeza) [inline account_analytic_tag/readme/DESCRIPTION.rst]:
```suggestion
This module restores the *account analytic tags* as a method for categorizing analytic entries, selectable from the invoices, and transferred to them on publishing.
```

File: account_analytic_tag/readme/CREDITS.rst
Comment (pedrobaeza) [inline account_analytic_tag/readme/CREDITS.rst:1]: Don't add empty files.

File: account_analytic_tag/readme/USAGE.rst
Comment (pedrobaeza) [inline account_analytic_tag/readme/USAGE.rst:1]: Please elaborate this a bit more, and put in `CONFIG.rst` the part that belongs there.

File: account_analytic_tag/readme/CONTRIBUTORS.rst
Comment (pedrobaeza) [inline account_analytic_tag/readme/CONTRIBUTORS.rst]: This is not the form we put it:

```suggestion
* `Tecnativa <https://www.tecnativa.com>`__:

  * Yadier Quesada
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
This is missing a lot of things:

- Analytic tags on invoices.
- Transfer such tags on the created analytic entries when publishing the invoices.

**Comment by Yadier-Tecnativa** [comment]
All suggestions were made, please review @pedrobaeza

**Comment by Yadier-Tecnativa** [comment]
ping @pilarvargas-tecnativa , @carlosdauden Please review

**Comment by rafaelbn** [comment]
Thank you! Great contribution! I will review this Friday!

**Comment by pepetreshere** [comment]
@Yadier-Tecnativa @pedrobaeza nice PR!, would you please drop some light on why do you remove the analytic_distribution_ids? The analytic_distribution_ids allowed to have a "predefined" percentage for each analytic account so that it is easier for and end-user to quickly select a set of "predefined" percentages. This is not covered by standard Odoo in v16.0 not even by using the new analytic plans. Thanks in advance for any light

**Comment by pedrobaeza** [comment]
Because this feature is already covered by standard v16.

**Comment by rafaelbn** [comment]
Hello! Really thank you very much for this contribution.

The implementation with this PR of analytic tags is different from Odoo 15 and previous. In this moment functionality of this module duplicated the function of Analytic Plans.

-------------------

- [ ] Analytic tags cannot have analytic account mandatory as they are independent, in fact don't need any relation.

You create analytic tags for many analytic accounts that will be created in the future (like projects)

Important: When the mind behind this decision in Odoo deleted the tags, he/she thought that all Odoo customers worldwide where using tags for distributions! But not, just tag. See

imagen

I have recorded a video and created an excel 😄 see:

- https://github.com/OCA/account-analytic/issues/551#issuecomment-1672273501

This PR should be able to reproduce that use case like in Odoo 15, 14, 13, 12, 11, etc.

--------------------

- [ ] There is not menu to add, see or edit analytic tags

imagen

---------------------

- [ ] The setting option to activate is missing

imagen


----------------------

- [ ] We cannot set analytic tags in invoice lines, the field is not available

----------------------


Let me know please if I can help you further

Regards,
Rafael

File: account_analytic_tag/README.rst
Comment (rafaelbn) [inline account_analytic_tag/README.rst]: @Yadier-Tecnativa this is not the function of analytic tags

File: account_analytic_tag/README.rst
Comment (rafaelbn) [inline account_analytic_tag/README.rst]: As commented previously not this is not. Please check the video I recorded 🙏🏼 😄

**Review by rafaelbn — CHANGES_REQUESTED** [review]

**Comment by sbidoul** [comment]
Thanks for this work. 

I think there are many different ways people were using analytic tags with other versions.

Would this make sense to you that we split the feature with first a base module that just provides the `account.analytic.tag` model, then additional modules that make use of them. Such as using tags on journal items, or settings tags on analytic accounts (as was possible until v11) ?

File: account_analytic_tag/models/account_analytic_tag.py
Comment (sbidoul) [inline account_analytic_tag/models/account_analytic_tag.py:22]: This m2o field seems wrong ?

**Comment by pedrobaeza** [comment]
> Would this make sense to you that we split the feature with first a base module that just provides the account.analytic.tag model, then additional modules that make use of them. Such as using tags on journal items, or settings tags on analytic accounts (as was possible until v11) ?

I don't see many sense on splitting model definition from account.move.line links at least. No need to extra dependencies (all are in `account`). It's complicating the module tree, even increasing maintenance burden, while the features are code are very simple to be maintained even if packed.

**Comment by pedrobaeza** [comment]
FYI, I will overtake this PR and finish it.

**Comment by sbidoul** [comment]
> I don't see many sense on splitting model definition from account.move.line links at least

In our use case we need the tags but not the link with `account.move.line` because the tags are set on analytic accounts (it was possible until v11). 

So I don't mind if the link with move lines is in the same module, but then we also need the group to show the analytic tags on move lines only to selected users.

In general I prefer small modules with restricted scope because they are easier to migrate. But I'm not blocking on this, especially if you do the work :).

> FYI, I will overtake this PR and finish it.

@pedrobaeza thanks! Do you know approximately when you'll work on that?

**Comment by pedrobaeza** [comment]
Probably next week.

**Comment by rafaelbn** [comment]
> I don't see many sense on splitting model definition from account.move.line links at least. No need to extra dependencies (all are in account). It's complicating the module tree, even increasing maintenance burden, while the features are code are very simple to be maintained even if packed.

I agree 👍🏼 , thank you!


> In our use case we need the tags but not the link with `account.move.line` because the tags are set on analytic accounts (it was possible until v11).

@sbidoul then you set `account.analytic.tag` in `account.analytic.account` and you don't want that this tags are propagated to `account.move.line` ? I don't catch it. Maybe we can talk about this in other moment (for me it is interesting)


> but then we also need the group to show the analytic tags on move lines only to selected users.

Interesting too. I guess the users with permissions in analytic with an extra security group can.


> FYI, I will overtake this PR and finish it.

Awesome!

Thank you ver much.

**Comment by victoralmau** [comment]
The following changes have been done:

- [x] Added checkbox at the billing configuration level for analytical tags.
- [x] Removed the active / inactive check in the labels.
- [x] Color field added in tree and form view.
- [x] Define the corresponding groups in the `company_id` field.
- [x] Analytic account field is not mandatory now to create labels.
- [x] The process of creating analytical items is simplified; now only analytical tags are set in the corresponding analytical items (those with the same or no analytical account defined in the label).
- [x] Create base tests to be able to use them in other modules that extend this functionality.
- [x] Add more use cases in tests.

File: account_analytic_tag/README.rst
Comment (rafaelbn) [inline account_analytic_tag/README.rst:40]: Go to Accounting -> Configuraton -> Settings -> Activate "Analytic Tags"

File: account_analytic_tag/models/account_analytic_tag.py
Comment (sbidoul) [reply account_analytic_tag/models/account_analytic_tag.py:22]: Can you add a help text that explains the purpose of this `account_analytic_id` field? Perhaps a more specific label than `Analytic Account` would help too.

File: account_analytic_tag/models/account_move_line.py
Comment (sbidoul) [inline account_analytic_tag/models/account_move_line.py]:
```suggestion
    analytic_tag_ids = fields.Many2many(
        comodel_name="account.analytic.tag",
        string="Analytic Tags",
    )
```

File: account_analytic_tag/views/account_analytic_line_views.xml
Comment (sbidoul) [inline account_analytic_tag/views/account_analytic_line_views.xml]: This `product_uom_category_id` field seems unrelated to this module ?

File: account_analytic_tag/views/account_analytic_line_views.xml
Comment (sbidoul) [inline account_analytic_tag/views/account_analytic_line_views.xml]: Can you make this field visible only to the `account_analytic_tag.group_analytic_tags` group?

File: account_analytic_tag/views/account_analytic_line_views.xml
Comment (sbidoul) [inline account_analytic_tag/views/account_analytic_line_views.xml]: Make it visible to the `account_analytic_tag.group_analytic_tags` group only.

File: account_analytic_tag/views/account_analytic_tag_views.xml
Comment (sbidoul) [inline account_analytic_tag/views/account_analytic_tag_views.xml]: Can we make this menu visible to the `account.group_account_manager` group instead. 

So accounting administrators can configure tags, but only users in the `account_analytic_tag.group_analytic_tags` can use them. This is important so we can use analytic tags just to categorize analytic accounts (in #585) but not show analytic tags on invoice/move lines.

**Review by sbidoul — COMMENTED** [review]
Thanks for this work. Here are a few comments.

**Comment by victoralmau** [comment]
> Thanks for this work. Here are a few comments.

Changes done.

**Comment by sbidoul** [comment]
Thanks! Code looks good to me. 

Do you know why the tag colors don't show on the invoice screen?

http://oca-account-analytic-16-0-pr572-a0742ba028bb.runboat.odoo-community.org/web?debug=1#id=18&cids=1&menu_id=115&action=232&active_id=1&model=account.move&view_type=form

image

image

File: account_analytic_tag/security/ir.model.access.csv
Comment (sbidoul) [inline account_analytic_tag/security/ir.model.access.csv]: @victoralmau oh, something needs fixing here too.

It should be readonly for `account_analytic_tag.group_analytic_tags` (not `analytic.group_analytic_accounting`, BTW).
And 1,1,1,1 for `account.group_account_manager` since managing the available tags is an accounting administration feature.

**Comment by victoralmau** [comment]
Changes done.

**Review by sbidoul — APPROVED** [review]
Thank you so much @victoralmau !

**Comment by rafaelbn** [comment]
I will finish my review **today**! This evening 😄

**Comment by pedrobaeza** [comment]
I'm thinking on reintroducing the distributions on tags as others have commented, as a way of "templating" this without defining a distribution model. What do you think about it?

**Comment by victoralmau** [comment]
> I'm thinking on reintroducing the distributions on tags as others have commented, as a way of "templating" this without defining a distribution model. What do you think about it?

This is a very good point.

Previously (up to v15) we all modeled the analytical distribution with labels, but now in v16 the concept of distributions has been introduced.

Advantages / disadvantages of each option.

**Option A: Keep the distributions in the labels**.

- [x] Disadvantage: Would duplicate functionality with distributions.
- [x] Advantage: It would work the same as in v15 and before.
- [x] Advantage: It will not be necessary to define a distribution on the invoice lines.
- [x] Disadvantage: It will be necessary to create the analytical entries when there is no distribution defined. What happens if there are labels and analytical distribution?
- [x] Advantage: No need to define distributions in OCA modules that add analytics (e.g. `project_stock`).

**Option B: Leave the analytical tags without distributions**.

- [x] Disadvantage: Having to define the concept of distributions in OCA modules that add analytics (e.g. `project_stock`).
- [x] Advantage: Not having to add code to create analytical entries without distributions.

Maybe option A is the most reasonable, what do others think?

**Comment by sbidoul** [comment]
> I'm thinking on reintroducing the distributions on tags as others have commented, as a way of "templating" this without defining a distribution model. What do you think about it?

@pedrobaeza  Can this be done in another module, or in a second step after merging this PR?

**Comment by sbidoul** [comment]
> Disadvantage: Would duplicate functionality with distributions

To me this justifies doing another analytic distributions mechanisms in a different module.

**Comment by victoralmau** [comment]
> > Disadvantage: Would duplicate functionality with distributions
> 
> To me this justifies doing another analytic distributions mechanisms in a different module.

Ok, thinking about this again, I think the best thing to do would be:

`account_analytic_tag_distribution`

- A checkbox is added in Invoicing > Configuration to enable distribution at analytic tag level.
- It hides the Analytic Distribution + Analytic Distribution columns menu if the checkbox is checked.
- The possibility to define the analytical distribution on tags is "restored".
- When defining an analytical label on an invoice line, what is defined on the labels will be defined as distribution.

In this way, I think it avoids "duplicating" the operation of the analytical distributions and could be used in one way or another.

**Review by pedrobaeza — APPROVED** [review]
OK about a separated module for distributions.

**Comment by rafaelbn** [comment]
Hello,

Thank you for this great work @victoralmau 

After review, I see this (let me know your opinions):

- [ ] In Invoice lines, set Analytic Tags after Analytic

imagen



- [ ] Go to Accounting -> Accounting -> Journal Items , Then we cannot group by "Analytic Account"

imagen



- [ ] Go to Accounting -> Accounting -> Management -> Analytic Items, change label "Tags" by "Analytic Tags" for better fields order (as alphabetical) and understanding

imagen


In a separate comment I can give my personal opinion about distributions.

Thank you very much! ❤️

**Review by rafaelbn — APPROVED** [review]
Hello,

Thank you for this great work @victoralmau 

After review, I see this (let me know your opinions):

- [x] In Invoice lines, set Analytic Tags after Analytic

imagen



- [ ] Go to Accounting -> Accounting -> Journal Items , Then we cannot group by "Analytic Account"

imagen



- [ ] Go to Accounting -> Accounting -> Management -> Analytic Items, change label "Tags" by "Analytic Tags" for better fields order (as alphabetical) and understanding

imagen


In a separate comment I can give my personal opinion about distributions.

Thank you very much! ❤️

**Comment by rafaelbn** [comment]
Hello, this is my opinion about a new module for distributions.

The most important writing this down here is that I don't want to be right, I am aware that there are many different types of Odoo implementations and different types of companies and clients. I don't want to convince anyone. I just want to provide a vision that may help someone else make their own decisions. In fact, **I really like reading your opinions and decisions** and they make me wonder why and what I can learn from it.

So...

If we in OCA reintroduce analytic distributions in tags with are confusing users further. As for every one in Odoo distributions are done in Analytic Plans (`account.analytic.plan`). This re-introducing feature for me, maybe could be decision prompted by the effort to understand the change and the technical dependencies of other modules.

I'm not opposed to anything 😄, this is an open community, but I do think that in addition to having a **separate module** we should make a separate repository like https://github.com/OCA/account-analytic-tags-distribution . Because we must keep an open scenario where contributors extend Odoo using Odoo analytics distributions with their analytics Plans.

In other case, this will be really confusing when developing modules around distributions in Analytic Plans like Odoo Core and Analytic Tags as older versions. In Odoo 18 nobody (new contributors) will remember about why are Distributions in Tags talking around product vision.

You have done a really nice analysis 👍🏼 take this point of view just if it gives you more perspective:

Perfect migrations from Odoo 15 to Odoo 16 for me is like:

- If `account.analytic.tag` has fields distribution "False" Then maintain it in `account.analytic.tag`
- If `account.analytic.tag` has fields distribution "True" Then migrate it to `account.analytic.plan`


Opining and introducing more points about this **great** 🥇 pros and cons analysis:

> Advantages / disadvantages of each option.
> 
> **Option A: Keep the distributions in the labels.**
> 
> - [x] Disadvantage: Would duplicate functionality with distributions.
> - [ ] Advantage: **It would work the same as in v15 and before.**

For me this is a **disadvantage** as you are separating important concepts in Odoo Core. Imagine you decide in Odoo 13 to maintain account.invoice model as a similar example.

> - [ ] Advantage: It will not be necessary to define a distribution on the invoice lines.
> - [x] Disadvantage: It will be necessary to create the analytical entries when there is no distribution defined. What happens if there are labels and analytical distribution?
> - [ ] Advantage: No need to define distributions in OCA modules that add analytics (e.g. project_stock).

For me this is not a Advantage as **you don't need to** define ~~distributions~~ analytic plans. The is always one by default that works in the same way as I you have in Odoo 15 only analytic accounts without distributions. Really I don't see the problem with `project_stock`
Are you talking around UX introducing invoices? I would really love to go deep in this problem that I don't see.

- [x] Disadvantage: You are missing the amazing new great functionality around **Applicability** in Analytic Plans with all the automation rules. https://www.odoo.com/documentation/16.0/applications/finance/accounting/reporting/analytic_accounting.html#analytic-accounting-analytic-plans

> **Option B: Leave the analytical tags without distributions.**
> 
> - [ ] Disadvantage: Having to define the concept of distributions in OCA modules that add analytics (e.g. project_stock).

As said before, you don't have to define concept of distributions. It's not a must, it is an option. Analytic Plans are just plans of analytic account, grouped, that maybe you you may want to distribute among them or not.

> 
> - [x] Advantage: Not having to add code to create analytical entries without distributions.
> 
> Maybe option A is the most reasonable, what do others think?

If you think is needed is OK. I think all the time about have we have (a lot) to not repeat my self (DRY)

1. "Analytic Distribution Models" `account.analytic.distribution.model` are not mandatory as they are rules for automations similar to previous "Analytic Defaults Rules" `account.analytic.default`
2. "Analytic Plans" `account.analytic.plan` are like "Dimensions", they group Analytic accounts and give you the possibility to make distributions or not.
3. Analytic Accounts, are values of the Plans (dimensions) with a direct relations
4. Analytic Tags `account.analytic.tag` in this PR are another level of analysis, for other values and dimensions that are not related so you can assign freely.


My best!, sorry for the length! (my fault)

💯 🥇 👍🏼 

Thank you very much for this work! 👍🏼

**Comment by victoralmau** [comment]
Changes done:
- [x]  In Invoice lines, set Analytic Tags after Analytic
- [x] Change `tag_ids` position in analytical items (same as in v15).

Other comments:
- There is no analytical account field in the Journal Items, they are built according to the analytical distribution.
- Tags is the correct name of the field (same as in v15) in the analytical items.
- The configuration section of the readme already indicates the activation of the "Analytic Tag" checkbox.

**Comment by pedrobaeza** [comment]
@victoralmau put `Co-Authored-By: ...` in the commit for your co-authorship.

**Comment by rafaelbn** [comment]
/ocabot merge nobump

Fantastic! You are awesome! 💚

**Pull request merged**
